### PR TITLE
Support for Active Directory multidomain forest.

### PR DESCRIPTION
**Pull request by nE0sIghT**
In AD forest samAccountName (or uid) may not be unique in the entire forest and userPrincipalName contains "@" symbol disallowed in Matrix User Identifiers.
This commit adds mxid generation logic to work around above restrictions.

Please note that ```exactly``` this commit was not yet tested by me with real AD forest, but very same implementation was tested.
Exactly this change will be tested by me next week.

### CODE REVIEW

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:97]: please could you add some comments here? what do each of these variables do, and what is the difference between `localpart` and `uid_value` ?

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:106]: a tuple is more conventional here:

```suggestion
                (login, domain) = localpart.rsplit(self.ldap_separator, 1)
```

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:104]: it would probably be a good idea to log something here, otherwise people are going to be confused when their ldap logins fail silently.

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:109]: could you factor all this out to a separate method? I think it would be clearer that way.

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:274]: why is the default domain special-cased here? why is it correct to strip the domain for accounts on that domain but not others?

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:390]: why is this restriction needed?

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:268]:
```suggestion
                (login, domain) = localpart.rsplit("@", 1)
```

File: README.rst
Comment (richvdh) [inline README.rst:82]: I suggest a new section here:

```suggestion

Active Directory forest support
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

File: README.rst
Comment (richvdh) [inline README.rst:87]: Indentifiers -> Identifiers

File: README.rst
Comment (richvdh) [inline README.rst:107]:
```suggestion
        # Optional. Specifies the domain separator character that users should enter. Defaults to `/`.
        mxid_domain_separator: /
```

File: README.rst
Comment (richvdh) [inline README.rst:90]: I think this could be reworded a bit.

```suggestion
If the ``ad_forest`` flag is set to ``true``, an Active Directory forest will be
searched for the login details.

In this mode, the user enters their login details in the form ``<login>/<domain>``. (The 
normal domain separators, ``@`` and ``\\``, cannot be used in Matrix User Identifiers, so 
``/`` is used by default instead.)
```

File: README.rst
Comment (richvdh) [inline README.rst:118]:
```suggestion
With this configuration the user can log in with either ``someuser/main.example.com`` or ``someuser``.
```

etc

**Comment by richvdh** [comment]
something else I forgot to say: we'll need sign-off from you to confirm that you are happy to contribute this code under the apache license. Sorry it's not spelt out more clearly, but please could you add a `Signed-off-by` to this PR as per https://github.com/matrix-org/synapse/blob/develop/CONTRIBUTING.md#sign-off ?

**Comment by nE0sIghT** [comment]
> why not allow the user to enter <user>\<domain> in that form, and then map that to the matrix user ID @<user>/<domain>:<server_name> ?

Thanks for suggestion. I thinked about it but missed the fact that authentication plugin may return any mxid.

**Comment by nE0sIghT** [comment]
> why not allow the user to enter <domain> in that form, and then map that to the matrix user ID @/:<server_name> ?

I reviewed [api documentation](https://github.com/matrix-org/synapse/blob/83434df3812650f53c60e91fb23c2079db0fb5b8/docs/password_auth_providers.md) and [Synapse source code](https://github.com/matrix-org/synapse/blob/83434df3812650f53c60e91fb23c2079db0fb5b8/synapse/handlers/auth.py#L723) and found that this is impossible with ```check_password``` method.  
Do you think we should switch to ```get_supported_login_types```/```check_auth``` pair from ```check_password```?

**Comment by richvdh** [comment]
sorry for being so slow to reply - this had slipped off my radar.

>  Do you think we should switch to get_supported_login_types/check_auth pair from check_password?

I think that might be a good idea, yes.

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:274]: 2 thoughts here:

1. This is very common pattern in corporate environment.  
You can look eg to [smb.conf's winbind use default domain](https://www.samba.org/samba/docs/current/man-html/smb.conf.5.html#idm10651).
Yes, it may be not recommended, bu it is widelly used.

2. Upgrading from previous versions of ```ldap_auth_provider``` while using ```samAccountName``` as uid.
After this PR merged it will be possible to continue using ```samAccountName``` as matrix uid for default domain users but also allow to use ```userPrincipalName``` for users of other domains in forest.

P.S.: ```samAccountName``` is first part of ```userPrincipalName``` before ```@```

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:390]: Those 2 symbols are only symbols allowed in mxid, but disallowed in ```userPrincipalName```.  
However let's drop this configuration in favor of binding ```/``` in mxid and using classic ```domain\login``` form as you suggested.

**Comment by richvdh** [comment]
please could you not force-push to git? it makes it much harder for us to see what has changed since the review.

**Comment by nE0sIghT** [comment]
Ah, sorry for that.  
Do you want me to restore 1st commit and add changes as new commit?  

I still don't fixed tests, will do it today.

**Comment by richvdh** [comment]
>  Do you want me to restore 1st commit and add changes as new commit?

that would be helpful, thanks!

**Comment by nE0sIghT** [comment]
> that would be helpful, thanks!

Done. I restored original commit, but added ```Signed-off-by``` tag.  
I will squash commits after review if needed.

Thanks for your review!

File: ldap_auth_provider.py
Comment (richvdh) [reply ldap_auth_provider.py:101]: if password is empty, then it should be handled the same way as any other incorrect password?

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:51]: it's not a matrix id.

```suggestion
    """Raised in case the user's login credentials cannot be mapped to a UPN"""
```

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:607]:
```suggestion
                    'No LDAP separator "/" was found in uid "%s" '
```

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:609]:
```suggestion
```

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:612]: seems like it might be simpler to just `return None` here, but :man_shrugging:

File: ldap_auth_provider.py
Comment (richvdh) [reply ldap_auth_provider.py:401]: ok, but it seems like this ought to be handled via documentation, rather than a hardcoded check. people may have systems that behave similarly to AD but expose a UPN-like attribute with a different name: there's no reason not to support that.

**Review by richvdh — CHANGES_REQUESTED** [review]
getting there, thank you! a few more comments

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:612]: If we will use ```None``` for that, then we can not just deserialize method output like

```
(login, domain, localpart) = self._map_login_to_upn(username)
```

So we will need some temporal variable like:
```
map_array = self._map_login_to_upn(username)
if not map_array:
    error here
(login, domain, localpart) = map_array
```

If you think it's better suits this project then I will refactor code. But I believe exception is better choice here.

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:101]: Sounds reasonable :-)

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:282]: Added "@" check since we dropped ```userPrincipalName``` restriction.

**Comment by nE0sIghT** [comment]
I added more commits. Hope they will address remaining issues.
Thanks!

File: README.rst
Comment (richvdh) [inline README.rst:113]: this is now incorrect (or at least misleading)?

File: README.rst
Comment (richvdh) [inline README.rst:126]:
```suggestion
Please note that ``userPrincipalName`` or a similar-looking LDAP attribute in the format
```

File: README.rst
Comment (richvdh) [inline README.rst:127]:
```suggestion
``login@domain`` must be used when the ``active_directory`` option is enabled.
```

File: ldap_auth_provider.py
Comment (richvdh) [inline ldap_auth_provider.py:133]: I'm still a bit surprised by this. why do we need to special-case it? why is an empty password any different to any other incorrect password?

File: ldap_auth_provider.py
Comment (richvdh) [reply ldap_auth_provider.py:607]: fair enough.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: ldap_auth_provider.py
Comment (nE0sIghT) [reply ldap_auth_provider.py:133]: That's how we could introduce security issue but thanks to ldap3 we don't.  
Long story short LDAP can not authenticate user without password. Attempting to bind user without password leads to anonymous bind which always success.  
Fortunately in ldap3 anonymous bind was splitted off to special case.  

I will revert empty password changes.

**Comment by nE0sIghT** [comment]
I added more commits for remaining issues and added tests for new functionality as well.  
I also fixed connection leak in ```check_3pid_auth``` which prevents doing tests with that method.

**Review by richvdh — APPROVED** [review]
lgtm. Thanks very much for all your work!

**Pull request merged**
